[export] provide refine function for automatically accepting dynamic shapes suggested fixes

pianpwk opened the pull request:
Summary:
Part of the work helping export's automatic dynamic shapes / dynamic shapes refining based on suggested fixes. 

Introduces a util function refine_dynamic_shapes_from_suggested_fixes() that takes the error message from a ConstraintViolationError message containing suggested dynamic shapes fixes, along with the original dynamic shapes spec, and returns the new spec. Written so that the suggested fixes from export can be directly parsed and used.

Example usage for the automatic dynamic shapes workflow:
```
# export, fail, parse & refine suggested fixes, re-export
try:
    export(model, inps, dynamic_shapes=dynamic_shapes)
except torch._dynamo.exc.UserError as exc:
    new_shapes = refine_dynamic_shapes_from_suggested_fixes(exc.msg, dynamic_shapes)
    export(model, inps, dynamic_shapes=new_shapes)
```

For examples of behavior, see the added test and docstring. Will take suggestions for renaming the function to something else 😅

Test Plan: test_export tests

Differential Revision: D57409142




cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @LucasLLC

angelayi commented on torch/export/dynamic_shapes.py:
```diff
@@ -896,3 +903,182 @@ def assoc_shape(t, dynamic_shape):
+
+    # check keys are existing dims or new roots
+    for k, c in suggested_fixes.items():
+        if not (k in name_to_dim or k in roots):
```
dumb question, what's the difference between `name_to_dim` and `roots`? it also seems like `roots` isn't used anywhere after this

pianpwk replied:
Oh not dumb.

name_to_dim tracks the original dynamic shapes, so if key is in there, it's a refinement of an original shape.

roots also contains introduced roots (sometimes we suggest adding a new dim as a root), so then it's a new dim.

If the key is in neither, then we don't know what to do with it.

I'll rewrite this to be less user facing probably.

pianpwk commented on torch/export/dynamic_shapes.py:
```diff
@@ -896,3 +903,182 @@ def assoc_shape(t, dynamic_shape):
+                ),
+            )
+        if isinstance(c, str):  # check dim/derived dim expression
+            expr = sympy.sympify(c)
```
Ah yep that's dumb coding.

pianpwk commented on torch/export/dynamic_shapes.py:
```diff
@@ -896,3 +903,182 @@ def assoc_shape(t, dynamic_shape):
+        if not isinstance(
+            c, (int, _Dim, _DerivedDim, str)
+        ):  # check derived dim expression
+            raise UserError(
+                UserErrorType.INVALID_INPUT,
+                (
+                    "Dictionary values to refine_with_suggested_fixes() must be integers, Dims, DerivedDims, "
+                    f"or strings representing derived dims, but instead found {c} of type {type(c)}"
+                ),
+            )
```
Right, sorry. At first this was split into two user-facing functions, so I'll change it.

avikchaudhuri commented on torch/export/dynamic_shapes.py:
```diff
@@ -896,3 +903,143 @@ def assoc_shape(t, dynamic_shape):
+    return name_to_dim
+
+
+def parse_and_refine_suggested_fixes(
```
Maybe `refine_dynamic_shapes_from_suggested_fixes`?

avikchaudhuri reviewed (approved):
We need to be a bit careful about saying `Suggested fixes` are only for dynamic shape refinements. E.g., we might have other kinds in the future (e.g., for data-dependent errors)

pianpwk commented:
> We need to be a bit careful about saying `Suggested fixes` are only for dynamic shape refinements. E.g., we might have other kinds in the future (e.g., for data-dependent errors)

This makes sense, I've changed the wording in places slightly.

pianpwk committed:
[export] provide parsing and refine function for automatically accepting dynamic shapes suggested fixes (#127436)

Summary:

Part of the work helping export's automatic dynamic shapes / dynamic shapes refining based on suggested fixes. 

Introduces a util function refine_dynamic_shapes_from_suggested_fixes() that takes the error message from a ConstraintViolationError message containing suggested dynamic shapes fixes, along with the original dynamic shapes spec, and returns the new spec. Written so that the dynamic shapes suggested fixes from export can be directly parsed and used.

Example usage for the automatic dynamic shapes workflow:
```
# export, fail, parse & refine suggested fixes, re-export
try:
    export(model, inps, dynamic_shapes=dynamic_shapes)
except torch._dynamo.exc.UserError as exc:
    new_shapes = refine_dynamic_shapes_from_suggested_fixes(exc.msg, dynamic_shapes)
    export(model, inps, dynamic_shapes=new_shapes)
```

For examples of behavior, see the added test and docstring. Will take suggestions for renaming the function to something else 😅

Potential followup: make this compatible with ShapesCollection

Test Plan: test_export tests

Reviewed By: avikchaudhuri

Differential Revision: D57409142
